### PR TITLE
ddl: schedule as many jobs as possible in one round, simplify sql to query jobs

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #54436

Problem Summary:

### What changed and how does it work?
- schedule as many jobs as possible in one round, simplify sql to query jobs
- replace 2 ddl test hook with failpoint
- rename methods

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

before this pr: create 100K tables under same database on owner on 1pd/3tidb/3tikv using 10 threads, run 2 time, takes 31m16/29m5. avg cpu about [335, 400]

after: run 3 times, takes 32m8/22m56/25m29, avg cpu about [200, 280]. the test result is not stable as expected, seems there are other components interfering DDL process, will check it later. but overall it it's better, both in time and resource usage

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/job_table.go [hunk 1]
```diff
@@ -360,6 +301,8 @@ func (s *jobScheduler) startDispatch() error {
 	// TODO move waitSchemaSyncedController out of ddlCtx.
 	s.clearOnceMap()
 	s.mustReloadSchemas()
+
+	logRateLimiter := rate.NewLimiter(rate.Every(30*time.Second), 3)
```
Comment (lance6716) [inline pkg/ddl/job_table.go:305]: Do you think we should use zapcore.NewSamplerWithOptions like this?

https://github.com/pingcap/tidb/blob/be16d490485c58fa902daf7b947b53f4b1b0f26e/pkg/statistics/handle/logutil/logutil.go#L37-L56

I can add it in a separate PR or in this PR.

**Comment by D3Hunter** [comment]
@lance6716

File: pkg/ddl/ddl_worker_test.go [hunk 2]
```diff
@@ -75,7 +73,7 @@ func TestAddBatchJobError(t *testing.T) {
 }
 
 func TestParallelDDL(t *testing.T) {
-	store, dom := testkit.CreateMockStoreAndDomainWithSchemaLease(t, testLease)
+	store, _ := testkit.CreateMockStoreAndDomainWithSchemaLease(t, testLease)
```
Comment (tangenta) [inline pkg/ddl/ddl_worker_test.go:76]:
```suggestion
	store := testkit.CreateMockStoreWithSchemaLease(t, testLease)
```

File: pkg/ddl/job_table.go [hunk 3]
```diff
@@ -431,33 +374,77 @@ func (s *jobScheduler) checkAndUpdateClusterState(needUpdate bool) error {
...
-		pool = s.reorgWorkerPool
-	}
-	wk, err := pool.get()
-	if err != nil || wk == nil {
-		logutil.DDLLogger().Debug(fmt.Sprintf("[ddl] no %v worker available now", pool.tp()), zap.Error(err))
-		return
+		involving := job.GetInvolvingSchemaInfo()
+		if targetPool.available() == 0 {
+			s.runningJobs.addPending(involving)
+			continue
```
Comment (tangenta) [inline pkg/ddl/job_table.go:411]: Why not block and wait?

**Review by tangenta — APPROVED** [review]

File: pkg/ddl/job_table.go (hunk 3 above)
Comment (D3Hunter) [reply pkg/ddl/job_table.go:411]: then DDL in another pool will wait much longer

File: pkg/ddl/job_table.go [hunk 4]
```diff
@@ -431,33 +374,77 @@ func (s *jobScheduler) checkAndUpdateClusterState(needUpdate bool) error {
...
-	if job == nil || err != nil {
+	defer s.runningJobs.resetAllPending()
+
+	const getJobSQL = `select reorg, job_meta from mysql.tidb_ddl_job %s order by job_id`
+	var whereClause string
+	if ids := s.runningJobs.allIDs(); len(ids) > 0 {
+		whereClause = fmt.Sprintf("where job_id not in (%s)", ids)
+	}
+	sql := fmt.Sprintf(getJobSQL, whereClause)
+	rows, err := se.Execute(context.Background(), sql, "load_ddl_jobs")
```
Comment (lance6716) [inline pkg/ddl/job_table.go:390]: use `s.schCtx`?

File: pkg/ddl/job_table.go [hunk 5]
```diff
@@ -431,33 +374,77 @@ func (s *jobScheduler) checkAndUpdateClusterState(needUpdate bool) error {
...
+
+		wk, err := targetPool.get()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		intest.Assert(wk != nil, "worker should not be nil")
+		if wk == nil {
+			// should not happen, we have checked available() before, and we are
+			// the only routine consumes worker.
+			logutil.DDLLogger().Info("no %v worker available now", zap.Stringer("type", targetPool.tp()))
```
Comment (lance6716) [inline pkg/ddl/job_table.go:436]: no need to use `%v`

File: pkg/ddl/job_table.go [hunk 6]
```diff
@@ -575,8 +564,11 @@ func (s *jobScheduler) delivery2Worker(wk *worker, pool *workerPool, job *model.
 	})
 }
 
-func (s *jobScheduler) runJobWithWorker(wk *worker, job *model.Job) error {
-	failpoint.InjectCall("beforeRunJobWithWorker")
+// runOneJobStep runs one step of the DDL job. we are using online-schema-change,
+// one job might go through multiple steps, each step is one job state change such
+// as from 'done' -> 'synced, or one schema state change such as 'delete only' -> 'write only'.
```
Comment (lance6716) [inline pkg/ddl/job_table.go:569]:
```suggestion
// as from 'done' -> 'synced', or one schema state change such as 'delete only' -> 'write only'.
```

File: pkg/ddl/job_table.go [hunk 7]
```diff
@@ -641,14 +633,6 @@ func (s *jobScheduler) runJobWithWorker(wk *worker, job *model.Job) error {
 	return nil
 }
 
-func (*jobScheduler) markJobProcessing(se *sess.Session, job *model.Job) error {
```
Comment (lance6716) [inline pkg/ddl/job_table.go:644]: Maybe involve another DDL dev to think about that we don't persist it.

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/job_table.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/ddl/job_table.go:390]: yes, this and below one are copied from previous impl

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**